raft/raft_test.go

```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
+
+	lead := nt.peers[1].(*raft)
+
+	// Transfer leader to isolated node, wait for timeout.
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgTransferLeader})
+	if lead.transferee == None {
+		t.Errorf("Waiting transferring, transferee = %v, want %v", lead.transferee, 3)
+	}
+	for i := 0; i < lead.heartbeatTimeout+1; i++ {
+		lead.tick()
+	}
+	if lead.transferee == None {
+		t.Errorf("Waiting transferring, transferee = %v, want %v", lead.transferee, 3)
+	}
+
+	for i := 0; i < lead.electionTimeout-lead.heartbeatTimeout+1; i++ {
+		lead.tick()
+	}
+
+	if lead.state != StateLeader || lead.lead != 1 || lead.transferee != None {
+		t.Fatalf("Node 1 has state %v lead %v transferee %v, want state %v leader %v transferee %v", lead.state, lead.lead, lead.transferee, StateLeader, 1, None)
+	}
+}
+
+func TestTransferIgnorePropose(t *testing.T) {
+	nt := newNetwork(nil, nil, nil)
+	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgHup})
+
+	nt.isolate(3)
+
+	lead := nt.peers[1].(*raft)
+
+	// Transfer leader to isolated node to let transfer pending, then send propose.
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgTransferLeader})
+	if lead.transferee == None {
+		t.Errorf("Waiting transferring, transferee = %v, want %v", lead.transferee, 3)
+	}
+
+	for j := 0; j <= 5; j++ {
+		nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgProp, Entries: []pb.Entry{{}}})
+	}
+
+	if lead.prs[1].Match != 1 {
+		t.Fatalf("Node 1 has match %x, want %x", lead.prs[1].Match, 1)
+	}
+}
+
+func TestTransferReceiveHighTermVote(t *testing.T) {
+	nt := newNetwork(nil, nil, nil)
+	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgHup})
+
+	nt.isolate(3)
+
+	lead := nt.peers[1].(*raft)
+
+	// Transfer leader to isolated node to let transfer pending.
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgTransferLeader})
+	if lead.transferee == None {
+		t.Errorf("Waiting transferring, transferee = %v, want %v", lead.transferee, 3)
+	}
+
+	nt.send(pb.Message{From: 2, To: 2, Type: pb.MsgHup, Index: 1, Term: 2})
+
+	if lead.state != StateFollower || lead.lead != 2 || lead.transferee != None {
+		t.Fatalf("Node 1 has state %v lead %v transferee %v, want state %v leader %v transferee %v", lead.state, lead.lead, lead.transferee, StateFollower, 2, None)
+	}
+}
+
+func TestTransferRemoveNode(t *testing.T) {
+	nt := newNetwork(nil, nil, nil)
+	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgHup})
+
+	nt.ignore(pb.MsgTimeoutNow)
+
+	lead := nt.peers[1].(*raft)
+
+	// The transferee is removed when lead transfering.
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgTransferLeader})
+	if lead.transferee == None {
+		t.Errorf("Waiting transferring, transferee = %v, want %v", lead.transferee, None)
+	}
+
+	lead.removeNode(3)
+
+	if lead.state != StateLeader || lead.lead != 1 || lead.transferee != None {
+		t.Fatalf("Node 1 has state %v lead %v transferee %v, want state %v leader %v transferee %v", lead.state, lead.lead, lead.transferee, StateLeader, 1, None)
+	}
+}
+
+func TestTransferBack(t *testing.T) {
```

Comment:
TestLeaderTransferBack.

we need to add comment to this test about what it does. it has two steps and it is not clear from the function name.